api/envoy/config/route/v3/route_components.proto

```diff
@@ -542,6 +542,34 @@ message RouteMatch {
     //
     // Expect the value to not contain ``?`` or ``#`` and not to end in ``/``
     string path_separated_prefix = 14 [(validate.rules).string = {pattern: "^[^?#]+[^?#/]$"}];
+
+    // If specified, the route is a template match rule meaning that the
+    // ``:path`` header (without the query string) must match the given
+    // ``path_template_match`` pattern.
+    //
+    // Path template matching types:
+    //
+    // * ``*`` : Matches a single path component, up to the next path separator: /
+    //
+    // * ``**`` : Matches zero or more path segments. If present, must be the last operator.
```

Comment:
Is there a formal grammar we can point people to as well? It doesn't have to be in the docs, but unlike somehting like RE2, this is not an open standard as such. @mattklein123 I'd like to confirm here that we want this as a 1st class API - I think this is a good idea as it's not regex and for the various reasons discussed historically. OTOH, it could theoretically be a pluggable regex engine (if you excuse the misnomer). The main challenge would be to include the substitution language as well I think.

Reply:
Friendly ping @mattklein123

Reply:
Sorry I missed this originally.

What are the reasons this is not regex? I assume performance and simplicity of implementation? Can we potentially spell out the reasons why someone would use this over regex because as you say regex with pattern matching could work also?

Also I agree that if this is some standard that is going to be shared across products it should be documented somewhere canonical?

Reply:
This is basically https://github.com/envoyproxy/envoy/issues/7763. The main thing I wanted to clarify is that, given the difference in syntax between regex and URL templates/globs, it would make more sense to treat it as first class rather than a pseudo-regex "engine.

Reply:
FWIW, I don't think it makes sense to treat this as a regex engine, since it's not regex syntax.  But I do wonder if it would make sense to add an extension point for a generic path-matching plugin here, and this could then be an extension that plugs into that extension point rather than a first-class mechanism baked directly into the API.  I'm not sure how many people will actually use this, and adding an extension point would also pave the way for future custom path-matching extensions when others want to do this differently.

Reply:
Yeah I'm fine either way between extension or not, but I think we should more clearly document the rationale behind why this exists and why one would use this over regex. Also, if this really is shared between other systems as I said before I think it would be better if there was some canonical documentation on the syntax.

Reply:
@mattklein123 and @htuch It sounds like the only action item here is additional docs. Are we ok to approve these set of changes?

Reply:
From my side, correct. As a tiny nit I might rename this field to `path_template`, but that's all API wise.

Reply:
I wanted to follow the same conversation as Media CDN where it's path_template_match. https://cloud.google.com/media-cdn/docs/routing#path-matching

If we're all set LGTM

Reply:
Prefer to preference Envoy API consistency here rather than a specific vendor product. LGTM with that change and the not-impelemnted-hide stuff we discussed.

Reply:
Changed name and added not-impelemnted-hide stuff. Should be all set